### PR TITLE
fix(#2109): floating help window

### LINKED ISSUE #2109: Mappings Executed In Help View
### Description

Mappings are executed in help mode. Fails most actions e.g. `a`

`E5108: Error executing lua: ...t/nvim-tree.lua/lua/nvim-tree/actions/fs/create-file.lua:34: attempt to index field 'absolute_path' (a nil value)`

### Neovim version

```text
NVIM v0.8.3
Build type: Release
LuaJIT 2.1.0-beta3
```

### Operating system and version

Linux  6.2.9-arch1

### nvim-tree version

d1410cb

### Minimal config

```lua
vim.cmd [[set runtimepath=$VIMRUNTIME]]
vim.cmd [[set packpath=/tmp/nvt-min/site]]
local package_root = "/tmp/nvt-min/site/pack"
local install_path = package_root .. "/packer/start/packer.nvim"
local function load_plugins()
  require("packer").startup {
    {
      "wbthomason/packer.nvim",
      "nvim-tree/nvim-tree.lua",
      "nvim-tree/nvim-web-devicons",
      -- ADD PLUGINS THAT ARE _NECESSARY_ FOR REPRODUCING THE ISSUE
    },
    config = {
      package_root = package_root,
      compile_path = install_path .. "/plugin/packer_compiled.lua",
      display = { non_interactive = true },
    },
  }
end
if vim.fn.isdirectory(install_path) == 0 then
  print "Installing nvim-tree and dependencies."
  vim.fn.system { "git", "clone", "--depth=1", "https://github.com/wbthomason/packer.nvim", install_path }
end
load_plugins()
require("packer").sync()
vim.cmd [[autocmd User PackerComplete ++once echo "Ready!" | lua setup()]]
vim.opt.termguicolors = true
vim.opt.cursorline = true

-- MODIFY NVIM-TREE SETTINGS THAT ARE _NECESSARY_ FOR REPRODUCING THE ISSUE
_G.setup = function()
  require("nvim-tree").setup {}
end
```

### Steps to reproduce

`g?`
`a`

### Expected behavior

No action.

### Actual behavior

`E5108: Error executing lua: ...t/nvim-tree.lua/lua/nvim-tree/actions/fs/create-file.lua:34: attempt to index field 'absolute_path' (a nil value)`

### PR DESCRIPTION
**Pull request by alex-courtis**
fixes #2109

Work in progress:

- [x] Styling
- [x] Remove length restrictions, formatting sanely
- [x] Remove old help plumbing / drawing
- [x] Simplify help.lua
- [x] More robust au
- [x] Compute mappings from API rather than current buffer
- [x] toggle
- [x] q exit

### CODE REVIEW

File: lua/nvim-tree/renderer/help.lua [hunk 1]
```diff
@@ -86,4 +88,50 @@ function M.compute_lines()
...
+    col = 1,
+    style = "minimal",
+    noautocmd = true,
+  })
+
+  -- style it a bit like the tree
+  vim.wo[winnr].cursorline = view.View.winopts.cursorline
+
+  -- close window and delete buffer on leave
+  vim.api.nvim_create_autocmd("WinLeave", {
```
Comment (gegoune) [inline lua/nvim-tree/renderer/help.lua:128]: Shouldn't it set once to true?

Comment (alex-courtis) [reply lua/nvim-tree/renderer/help.lua:128]: Thank you!

File: lua/nvim-tree/help.lua [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+
+  -- highlight it
+  for _, h in ipairs(hl) do
+    vim.api.nvim_buf_add_highlight(M.bufnr, -1, h[1], h[2], h[3], h[4])
+  end
+
+  -- open a very restricted window
+  M.winnr = vim.api.nvim_open_win(M.bufnr, true, {
+    relative = "editor",
+    border = "rounded",
```
Comment (gegoune) [inline lua/nvim-tree/help.lua:156]:
```suggestion
    border = "none",
```

File: lua/nvim-tree/help.lua [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+    width = width,
+    height = #lines,
+    row = 1,
+    col = 1,
```
Comment (gegoune) [inline lua/nvim-tree/help.lua:160]:
```suggestion
    col = 0,
```

File: lua/nvim-tree/help.lua [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+
+  -- longest lhs and description
+  local max_lhs = 0
+  local max_desc = 0
+  for _, l in pairs(mappings) do
+    max_lhs = math.max(#l.lhs, max_lhs)
+    max_desc = math.max(#l.desc, max_desc)
+  end
+
+  local fmt = string.format("%%-%ds %%-%ds", max_lhs, max_desc)
```
Comment (gegoune) [inline lua/nvim-tree/help.lua:107]:
```suggestion
  local fmt = string.format(" %%-%ds %%-%ds", max_lhs, max_desc)
```

File: lua/nvim-tree/help.lua [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+  -- remainder alpha
+  return a:gsub("[^a-zA-Z]", "") < b:gsub("[^a-zA-Z]", "")
+end
+
+--- Compute all lines for the buffer
+--- @return table strings of text
+--- @return table arrays of arguments 3-6 for nvim_buf_add_highlight()
+--- @return number maximum length of text
+local function compute()
+  local lines = { "nvim-tree mappings   exit: q" }
```
Comment (gegoune) [inline lua/nvim-tree/help.lua:85]:
```suggestion
```

File: lua/nvim-tree/help.lua [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+  end)
+
+  -- longest lhs and description
+  local max_lhs = 0
+  local max_desc = 0
+  for _, l in pairs(mappings) do
+    max_lhs = math.max(#l.lhs, max_lhs)
+    max_desc = math.max(#l.desc, max_desc)
+  end
+
```
Comment (gegoune) [inline lua/nvim-tree/help.lua:106]:
```suggestion

  local lines = { ("nvim-tree mappings%sexit: q"):format(string.rep(" ", max_desc + max_lhs - 23)) }
```

File: lua/nvim-tree/help.lua (hunk 5 above)
Comment (gegoune) [reply lua/nvim-tree/help.lua:85]: Moving down to where we know the width.

File: lua/nvim-tree/help.lua (hunk 6 above)
Comment (gegoune) [reply lua/nvim-tree/help.lua:106]: trying to right pad `exit: q`. Still seeing some space to the right but it's most likely just window border/window separator. Keen to see how it looks in your setup.
pad

File: lua/nvim-tree/help.lua (hunk 4 above)
Comment (gegoune) [reply lua/nvim-tree/help.lua:107]: Adding padding after removal of border and moving window all the way to the edge.

File: lua/nvim-tree/help.lua (hunk 2 above)
Comment (gegoune) [reply lua/nvim-tree/help.lua:156]: Opinion: look cleaner without border.

File: lua/nvim-tree/help.lua (hunk 3 above)
Comment (gegoune) [reply lua/nvim-tree/help.lua:160]: To cover over sign column.

But we really should check if tree is displayed on left or right side of screen here.

**Review by gegoune — CHANGES_REQUESTED** [review]

**Comment by alex-courtis** [comment]
Looking good! Many thanks.

Could you stretch to a single border? I was just updated to neovim 0.9 and they changed the default float background to dark:

**Comment by alex-courtis** [comment]
Please do just push to the branch in the future... I'm completely OK with cowboy commits.

**Pull request merged**